vshard/storage/schema.lua

```diff
@@ -256,7 +181,14 @@ local function schema_upgrade(target_version, username, password)
             if errinj == 'begin' then
                 ok, err1 = false, 'Errinj in begin'
             else
-                ok, err1 = pcall(handler.upgrade, username)
+                ok, err1 = pcall(function()
+                    local exports = schema_find_exports_for_vshard_version(
+                        next_version)
+                    exports = lvexports.compile(exports)
+                    handler.upgrade()
```

Comment:
Why are we trying to update `_schema` several lines below, when we have already done that in `handler.upgrade()`?

```lua
            ok, err1 = pcall(_schema.replace, _schema,
                             {'vshard_version', unpack(next_version)})
            if not ok then
                llog.info("Upgraded schema to %s but couldn't update _schema "..
                          "'vshard_version' - fatal error: '%s'", next_version,
                          err1)
                os.exit(-1)
            end
```

Reply:
> Why are we trying to update _schema several lines below, when we have already done that in handler.upgrade()?

In `schema_init_0_1_15_0()` it is set because this is an init-function, it is expected to create the complete schema, including the version.

In `schema_upgrade_to_0_1_16_0()` it is set, because we switch from the legacy version storage `'oncevshard:storage:1'` to `'vshard_version'`.

It might look excess, but if we would have, say, `schema_upgrade_to_0_1_25_0()`, then it wouldn't touch the version, because nothing about the version-storage would change. It is just not very clear right now because we have only 2 versions and they both need to touch the version as a part of their upgrade routine.

I do agree though, that we could drop the version setting from `schema_upgrade_to_0_1_16_0()` and probably nothing would change. I was thinking about it when touched this code and decided not to touch it, after all 😁.

Reply:
> It might look excess, but if we would have, say, schema_upgrade_to_0_1_25_0(), then it wouldn't touch the version, because nothing about the version-storage would change. It is just not very clear right now because we have only 2 versions and they both need to touch the version as a part of their upgrade routine.

I see now, thank you for the explanation